# 📚 Word2Vec Embeddings – Theoretical Overview

---

## 🔹 What is Word2Vec?

**Word2Vec** is a technique used to convert words into **dense vector representations**, also known as **word embeddings**.

These embeddings capture **semantic relationships** between words by placing similar words **closer together in a continuous vector space**.

For example:


---

## 🔹 Why Use Word Embeddings?

- Traditional methods (like one-hot encoding) treat each word as independent and unrelated.
- Word2Vec, however, captures **context and meaning**.
- It reduces **dimensionality** and preserves **semantic similarity**.

---

## 🔹 How Does Word2Vec Work?

Word2Vec is based on the idea that **a word is defined by the company it keeps** (distributional hypothesis).

There are two main architectures:

1. **CBOW (Continuous Bag of Words):**  
   - Predicts a word based on its **context words**.
   - Example: Predict `"attack"` from `["the", "on", "was"]`.

2. **Skip-Gram:**  
   - Predicts **context words** from a given word.
   - Example: Use `"attack"` to predict `"the"`, `"on"`, `"was"`.

Both methods use a shallow neural network to learn embeddings. The **hidden layer's weights become the word vectors**.

---

## 🔹 What Do the Embeddings Represent?

- Each word is represented by a **vector of real numbers** (e.g., 50 or 100 dimensions).
- These vectors encode:
  - **Semantic similarity** (e.g., `"terror"` close to `"violence"`)
  - **Syntactic roles**
  - **Analogical reasoning** (e.g., `"paris" - "france" + "italy" ≈ "rome"`)

---

## 🔹 Custom vs. Pretrained Word2Vec Embeddings

### ✅ Custom-Trained Embeddings

- You train Word2Vec on **your own dataset** (e.g., 50 terrorism-related sentences).
- Advantages:
  - Tailored to your domain and vocabulary.
  - Captures context specific to your data.
- Disadvantages:
  - Needs sufficient text data to be meaningful.
  - Poor generalization outside the training corpus.

### ✅ Pretrained Word2Vec Embeddings

- Trained on massive corpora like **Google News**, **Wikipedia**, or **Common Crawl**.
- Common sources:
  - `GoogleNews-vectors-negative300.bin` (3 million words, 300 dimensions)
- Advantages:
  - Very rich semantic information.
  - Ready to use and effective even with small datasets.
- Disadvantages:
  - May include vocabulary irrelevant to your task.
  - Large file sizes (can be hundreds of MBs).

---

## 🔹 Summary

| Feature           | Custom-Trained                | Pretrained                          |
|------------------|-------------------------------|-------------------------------------|
| Data Source       | Your own dataset              | Public corpora (Google News, etc.)  |
| Vocabulary        | Limited to your text          | Broad and extensive                 |
| Domain Fit        | Excellent for your domain     | General-purpose                     |
| Training Needed   | Yes                           | No                                  |
| Quality           | Varies (needs data)           | High                                |

---

Word2Vec is a powerful way to represent language numerically and has paved the way for many modern NLP techniques. Whether you use custom or pretrained embeddings depends on your data size, domain specificity, and task requirements.


In [ ]:
pip install gensim

In [ ]:
import pandas as pd
from gensim.models import Word2Vec
from nltk.tokenize import word_tokenize
import nltk
import numpy as np

nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [ ]:
# Example data
data = {'Text': [
    'The attack caused widespread panic',
    'Terrorists threatened national security'
]}
df = pd.DataFrame(data)
df

,Text
0,The attack caused widespread panic
1,Terrorists threatened national security


In [ ]:
import nltk
nltk.download('punkt_tab')

[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt_tab.zip.


True

## 🧠 Traditional Text Vectorization Techniques (Before Word2Vec)

Before deep learning-based embeddings like **Word2Vec**, text data was converted into numerical form using simpler, frequency-based techniques. These approaches did not capture semantic meaning but were widely used in early Natural Language Processing (NLP) tasks.

---

### 🔹 1. Bag of Words (BoW)

- Represents text as a **vector of word counts**.
- Ignores word order and context.
- The size of the vector equals the number of unique words in the corpus.
- Example:
  - Sentence: "Terrorists threatened security"
  - Vector: [1, 1, 1] → one count for each word.

**Pros:**
- Simple and fast.
- Easy to implement.

**Cons:**
- No information about word meaning or position.
- High dimensional and sparse.

---

### 🔹 2. TF-IDF (Term Frequency – Inverse Document Frequency)

- Improves BoW by reducing the weight of **common words** and increasing the importance of **rare but informative words**.
- Formula:
  - `TF-IDF = TF(w,d) * log(N / DF(w))`
- Results in better text representations than raw counts.

**Pros:**
- Reduces impact of frequently occurring but less informative words like "the", "is", etc.
- Still simple to compute.

**Cons:**
- Still ignores word order and semantics.
- Sparse representation.

---

### ❗ Limitations of BoW and TF-IDF

- **No semantic understanding** (e.g., "car" and "automobile" are unrelated numerically).
- **Context is lost** — can't distinguish different meanings of the same word.
- **High-dimensional vectors** lead to inefficiency and overfitting.

➡️ These limitations led to the development of **dense, semantic word embeddings** like **Word2Vec**, which map words to vectors in a way that captures their **meaning and context**.


In [ ]:
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
# ==========================
# Bag of Words (BoW)
# ==========================
bow_vectorizer = CountVectorizer()
bow_features = bow_vectorizer.fit_transform(df['Text'])

# Convert to DataFrame
bow_df = pd.DataFrame(bow_features.toarray(), columns=bow_vectorizer.get_feature_names_out())
print("🔹 Bag of Words Features:")
display(bow_df)

# ==========================
# TF-IDF
# ==========================
tfidf_vectorizer = TfidfVectorizer()
tfidf_features = tfidf_vectorizer.fit_transform(df['Text'])

# Convert to DataFrame
tfidf_df = pd.DataFrame(tfidf_features.toarray(), columns=tfidf_vectorizer.get_feature_names_out())
print("🔹 TF-IDF Features:")
display(tfidf_df)

🔹 Bag of Words Features:


,attack,caused,national,panic,security,terrorists,the,threatened,widespread
0,1,1,0,1,0,0,1,0,1
1,0,0,1,0,1,1,0,1,0


🔹 TF-IDF Features:


,attack,caused,national,panic,security,terrorists,the,threatened,widespread
0,0.447214,0.447214,0.0,0.447214,0.0,0.0,0.447214,0.0,0.447214
1,0.000000,0.000000,0.5,0.000000,0.5,0.5,0.000000,0.5,0.000000


# Tokenize and lowercase

In [ ]:
# Tokenize and lowercase
sentences = df['Text'].apply(lambda x: word_tokenize(x.lower())).tolist()
sentences

[['the', 'attack', 'caused', 'widespread', 'panic'],
 ['terrorists', 'threatened', 'national', 'security']]

# Word2Vec Custom Embeddings


## ⚙️ Word2Vec Training Parameters – Theoretical Explanation

---

### 🔹 `vector_size`

- Specifies the number of dimensions in which each word is represented.
- Determines the **size of the word embedding vector** (e.g., 5, 100, 300).
- A higher size can capture more complex semantic relationships but may require more data and training time.

---

### 🔹 `window`

- Sets the **context window size** — the number of words before and after the target word to consider as context.
- For example, if `window=3`, then for the word `"attack"`, the model will consider up to 3 words on either side.
- Smaller values focus on **local context**, while larger values capture **broader context**.

---

### 🔹 `min_count`

- Ignores all words with total frequency **lower than this threshold**.
- Helps filter out rare words that may add noise or be irrelevant.
- Useful for focusing the model on more meaningful and frequently occurring words.

---

### 🔹 `workers`

- Sets the number of **CPU threads** used during training.
- This allows parallelization and speeds up training on large datasets.

---

### 🔹 `sg`

- Determines the **training algorithm**:
  - `sg = 1`: Use **Skip-Gram** — predicts context words based on the current word.
  - `sg = 0`: Use **CBOW (Continuous Bag of Words)** — predicts the current word based on its context.
- **Skip-Gram** is better for learning rare words and works well for smaller datasets.
- **CBOW** is faster and effective on larger corpora.

---

These parameters guide how Word2Vec learns the relationships between words and builds dense vector representations that capture **semantic meaning**, **context**, and **word similarity**.


In [ ]:
# Train Word2Vec
model = Word2Vec(sentences, vector_size=5, window=3, min_count=1, workers=2, sg=1)

## 🧾 Printing Word Embeddings – Theoretical Explanation

---

### 🔹 Vocabulary in Word2Vec

Once a Word2Vec model is trained, it creates a **vocabulary** — a list of all words that met the `min_count` threshold during training.

Each of these words is assigned a **dense vector** (embedding) that captures its **semantic and contextual meaning**.

---

### 🔹 Accessing Embeddings

For every word in the vocabulary:
- The model stores a **corresponding vector** in its internal structure.
- This vector is a **real-valued array** (e.g., 5, 100, or 300 dimensions depending on `vector_size`).
- These vectors are stored in an ordered list called the **embedding matrix**.

---

### 🔹 Purpose of Printing Embeddings

When you loop through all words in the vocabulary and print their embeddings:
- You are **examining the learned representations** for each word.
- This helps you understand:
  - How the model perceives each word.
  - Which words have similar embeddings (i.e., similar meanings).
  - The structure of the **semantic space** created by the model.

---

### 🔹 Example Output Interpretation

For each word:
- You see a vector like `[0.12, -0.45, 0.33, ..., 0.08]`.
- These values **do not have human-readable meaning individually**.
- But together, they represent the **position of the word in vector space**, where distance and direction relate to meaning.

---

### 🔹 Practical Uses

- Visualizing embeddings (e.g., using t-SNE or PCA).
- Measuring **similarity** between words using cosine similarity.
- Using embeddings as **input features** for downstream NLP tasks (e.g., classification, clustering).

---

By printing all embeddings, you're effectively viewing the **knowledge Word2Vec has learned** about your dataset’s vocabulary.


In [ ]:
# Print embeddings for all words in the vocabulary
for word in model.wv.index_to_key:
    print(f"Word: '{word}'")
    print(f"Embedding: {model.wv[word]}\n")


Word: 'security'
Embedding: [-0.01072454  0.00472863  0.10206699  0.18018547 -0.186059  ]

Word: 'national'
Embedding: [-0.14233617  0.12917745  0.17945977 -0.10030856 -0.07526743]

Word: 'threatened'
Embedding: [ 0.1476101  -0.03066943 -0.09073226  0.13108103 -0.09720321]

Word: 'terrorists'
Embedding: [-0.03632035  0.0575316   0.01983747 -0.1657043  -0.18897636]

Word: 'panic'
Embedding: [0.14623532 0.10140524 0.13515386 0.01525731 0.12701781]

Word: 'widespread'
Embedding: [-0.06810732 -0.01892803  0.11537147 -0.15043275 -0.07872207]

Word: 'caused'
Embedding: [-0.15023164 -0.01860085  0.19076237 -0.14638333 -0.04667537]

Word: 'attack'
Embedding: [-0.03875482  0.16154873 -0.11861791  0.00090325 -0.09507468]

Word: 'the'
Embedding: [-0.192071    0.10014586 -0.17519172 -0.0878365  -0.000702  ]



## 📊 Understanding Sentence Embeddings from Word2Vec

---

### 🔹 Context

Once a Word2Vec model is trained, every word in the vocabulary is represented by a vector of real numbers (e.g., 5 values if `vector_size=5`).

To generate a **sentence embedding**, we combine the vectors of all words in the sentence — most commonly by computing the **average of their embeddings**.

---

### 🧩 Step-by-Step Example

Let’s analyze the following two sentences and their embeddings:

#### 🔸 Sentence 1:
**"The attack caused widespread panic"**

### 🔸 Step 1: Word Embeddings
Words and their embeddings:

| Word         | Embedding (vector of 5 values) |
|--------------|---------------------------------|
| the          | `[-0.192071, 0.10014586, -0.17519172, -0.0878365, -0.000702]`  
| attack       | `[-0.03875482, 0.16154873, -0.11861791, 0.00090325, -0.09507468]`  
| caused       | `[-0.15023164, -0.01860085, 0.19076237, -0.14638333, -0.04667537]`  
| widespread   | `[-0.06810732, -0.01892803, 0.11537147, -0.15043275, -0.07872207]`  
| panic        | `[0.14623532, 0.10140524, 0.13515386, 0.01525731, 0.12701781]`  

➡️ To compute the sentence embedding:
- Add all corresponding dimensions across the 5 vectors.
- Divide by the number of words (5 in this case).

### 🔸 Step 2: Add Embeddings Dimension-wise

We sum each dimension **across all 5 vectors**:

| Dimension | Sum of Values |
|-----------|----------------|
| 1st dim   | `-0.192071 + (-0.03875482) + (-0.15023164) + (-0.06810732) + 0.14623532 = -0.30392946`  
| 2nd dim   | `0.10014586 + 0.16154873 + (-0.01860085) + (-0.01892803) + 0.10140524 = 0.32557095`  
| 3rd dim   | `-0.17519172 + (-0.11861791) + 0.19076237 + 0.11537147 + 0.13515386 = 0.14747807`  
| 4th dim   | `-0.0878365 + 0.00090325 + (-0.14638333) + (-0.15043275) + 0.01525731 = -0.36849202`  
| 5th dim   | `-0.000702 + (-0.09507468) + (-0.04667537) + (-0.07872207) + 0.12701781 = -0.09415631`  

---

### 🔸 Step 3: Divide Each Sum by the Number of Words (5)

This gives the **average embedding** (sentence vector):

| Dimension | Calculation | Result |
|-----------|-------------|--------|
| 1st dim   | `-0.30392946 / 5` | `-0.06058589`  
| 2nd dim   | `0.32557095 / 5`  | `0.06511419`  
| 3rd dim   | `0.14747807 / 5`  | `0.02949562`  
| 4th dim   | `-0.36849202 / 5` | `-0.07369840`  
| 5th dim   | `-0.09415631 / 5` | `-0.01883126`  


📌 Final sentence embedding (mean of 5 word vectors):

`[-0.060585897, 0.06511419, 0.029495616, -0.07369841, -0.018831262]`

---

#### 🔸 Sentence 2:
**"Terrorists threatened national security"**

### 🔸 Step 1: Word Embeddings
Words and their embeddings:

| Word          | Embedding (vector of 5 values) |
|---------------|---------------------------------|
| terrorists    | `[-0.03632035, 0.0575316, 0.01983747, -0.1657043, -0.18897636]`  
| threatened    | `[0.1476101, -0.03066943, -0.09073226, 0.13108103, -0.09720321]`  
| national      | `[-0.14233617, 0.12917745, 0.17945977, -0.10030856, -0.07526743]`  
| security      | `[-0.01072454, 0.00472863, 0.10206699, 0.18018547, -0.186059]`  

### 🔸 Step 2: Add Embeddings Dimension-wise

We compute the sum of each corresponding dimension across all 4 word vectors:

| Dimension | Sum of Values |
|-----------|----------------|
| 1st dim   | `-0.03632035 + 0.1476101 + (-0.14233617) + (-0.01072454) = -0.04177096`  
| 2nd dim   | `0.0575316 + (-0.03066943) + 0.12917745 + 0.00472863 = 0.16076825`  
| 3rd dim   | `0.01983747 + (-0.09073226) + 0.17945977 + 0.10206699 = 0.210632`  
| 4th dim   | `-0.1657043 + 0.13108103 + (-0.10030856) + 0.18018547 = 0.04525364`  
| 5th dim   | `-0.18897636 + (-0.09720321) + (-0.07526743) + (-0.186059) = -0.547506`  

---

### 🔸 Step 3: Divide Each Sum by the Number of Words (4)

We compute the average of each dimension:

| Dimension | Calculation        | Result         |
|-----------|--------------------|----------------|
| 1st dim   | `-0.04177096 / 4`  | `-0.01044274`  
| 2nd dim   | `0.16076825 / 4`   | `0.04019206`  
| 3rd dim   | `0.210632 / 4`     | `0.05265799`  
| 4th dim   | `0.04525364 / 4`   | `0.01131341`  
| 5th dim   | `-0.547506 / 4`    | `-0.1368765`  

➡️ Final sentence embedding is the **average of the 4 word vectors**:

📌 Final sentence embedding:

`[-0.010442743, 0.040192064, 0.05265799, 0.011313409, -0.1368765]`

---

### 🧠 What Does This Mean?

- The resulting sentence vector captures the **combined semantic meaning** of the sentence.
- While individual numbers don't have standalone meaning, the position of the vector in high-dimensional space can be compared with other vectors to measure **similarity**, **clustering**, or **semantic distance**.

---

### 📌 Summary

- Each word is represented as a 5-dimensional vector learned by Word2Vec.
- To compute the sentence embedding, we take the **average of all word vectors** in the sentence.
- This aggregated vector represents the **semantic meaning of the entire sentence**, allowing it to be used in further NLP tasks such as classification, clustering, or similarity analysis.


In [ ]:
# Function to create sentence embedding
def get_sentence_embedding(tokens):
    vectors = [model.wv[token] for token in tokens if token in model.wv]
    return np.mean(vectors, axis=0) if vectors else np.zeros(model.vector_size)

# Apply to the DataFrame
df['embedding'] = [get_sentence_embedding(tokens) for tokens in sentences]


# Show results
df

,Text,embedding
0,The attack caused widespread panic,"[-0.060585897, 0.06511419, 0.029495616, -0.07369841, -0.018831262]"
1,Terrorists threatened national security,"[-0.010442743, 0.040192064, 0.05265799, 0.011313409, -0.1368765]"


In [ ]:
import pandas as pd
import numpy as np
import gensim.downloader as api
from nltk.tokenize import word_tokenize
import nltk
nltk.download('punkt')

# 1. Load pretrained Google Word2Vec model (takes time)
model = api.load("word2vec-google-news-300")  # 300-dimension vectors

# 2. Example sentences
data = {'Text': [
    'The attack caused widespread panic',
    'Terrorists threatened national security'
]}
df = pd.DataFrame(data)

# 3. Tokenize each sentence
df['tokens'] = df['Text'].apply(lambda x: word_tokenize(x.lower()))
df


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


[==================================================] 100.0% 1662.8/1662.8MB downloaded


,Text,tokens
0,The attack caused widespread panic,"[the, attack, caused, widespread, panic]"
1,Terrorists threatened national security,"[terrorists, threatened, national, security]"


In [ ]:
# 4. Function to get word embeddings
def get_word_embeddings(tokens):
    return {word: model[word] for word in tokens if word in model}

# 5. Function to get sentence embedding (mean of word vectors)
def get_sentence_embedding(tokens):
    vectors = [model[word] for word in tokens if word in model]
    return np.mean(vectors, axis=0) if vectors else np.zeros(model.vector_size)



In [ ]:
# 6. Apply to get embeddings
df['word_embeddings'] = df['tokens'].apply(get_word_embeddings)
df['sentence_embedding'] = df['tokens'].apply(get_sentence_embedding)

# 7. View results
pd.set_option('display.max_colwidth', None)
df

Text  \
0       The attack caused widespread panic   
1  Terrorists threatened national security   

                                         tokens  \
0      [the, attack, caused, widespread, panic]   
1  [terrorists, threatened, national, security]   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 

# Golve

In [ ]:
import gensim.downloader as api
glove_vectors = api.load("glove-wiki-gigaword-100")
print("Pretrained Vector for 'bomb':\n", glove_vectors['bomb'])

[==================================================] 100.0% 128.1/128.1MB downloaded
Pretrained Vector for 'bomb':
 [-0.51737   0.67008  -0.029651 -0.7386    0.14512   0.2893    0.31469
 -0.025834  0.68313   2.1249    0.82737   0.051467  0.27282   1.2407
  0.057461  0.90143  -0.069999  0.036999 -0.95994   0.051862  0.8849
 -0.72432   0.29531  -0.83296  -0.018305  0.18601  -0.38575   0.75082
 -0.18182   0.74373  -0.32542  -0.30949   0.055913  0.49645   0.14176
 -0.42837   0.37072  -0.46903   0.7006    0.56325  -0.12873   0.05763
  1.1156   -0.37415   0.66822   0.078093  0.40486   0.21578  -0.8398
 -0.54327  -0.17044   0.50108  -0.18644   1.1879   -0.74442  -1.9035
 -1.1216    0.49935   1.7585    0.096104  0.096139  1.5203    0.54279
  0.35675  -0.055548 -0.15721  -0.38882  -0.15273  -0.35632   0.21487
 -0.53805  -0.65829   0.26204  -0.26519  -0.26225   0.43873   1.4172
 -0.42517  -1.3216    0.52509   0.78635   0.061809 -0.64796  -0.45048
 -0.46517  -0.36998   0.66681   0.26963  -0.73925

# FastText

In [ ]:
fasttext_vectors = api.load("fasttext-wiki-news-subwords-300")
print("FastText Vector for 'explosives':\n", fasttext_vectors['explosives'])

[==================================================] 100.0% 958.5/958.4MB downloaded
FastText Vector for 'explosives':
 [-0.023993    0.0084899  -0.03401     0.0062023   0.046256   -0.043538
 -0.00050459 -0.051844    0.029446    0.062069    0.0028935  -0.0073904
  0.046168    0.040937   -0.016748   -0.0029417   0.028633    0.05261
  0.0045006   0.019085    0.01301     0.031823    0.04045     0.023192
 -0.014248    0.0045493  -0.017011   -0.0095423   0.12552     0.0089388
  0.024653   -0.05398     0.0072956  -0.021797    0.01489     0.011422
 -0.037883    0.035104   -0.07204    -0.017371   -0.047275   -0.070364
  0.087057   -0.041285    0.041032   -0.02438     0.0027797   0.055012
 -0.0093959   0.031346   -0.0081277  -0.0026273  -0.026284   -0.018927
 -0.071982    0.044864   -0.0052574   0.034481   -0.027093    0.0034242
 -0.0067233   0.0071427   0.094684    0.023461    0.043478    0.066417
  0.074095   -0.057202   -0.015529    0.0011741   0.094605   -0.030872
  0.059662   -0.021885   -

# BERT


In [ ]:
from transformers import BertTokenizer, BertModel
import torch

tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
model = BertModel.from_pretrained('bert-base-uncased')

text = "The terrorist planted explosives near the government building."
inputs = tokenizer(text, return_tensors='pt')
outputs = model(**inputs)

# Get the embeddings for [CLS] token
cls_embedding = outputs.last_hidden_state[0][0]
print("BERT [CLS] Embedding Shape:", cls_embedding.shape)
print("BERT [CLS] Embedding Vector:\n", cls_embedding.detach().numpy())

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/440M [00:00<?, ?B/s]

BERT [CLS] Embedding Shape: torch.Size([768])
BERT [CLS] Embedding Vector:
 [-1.98548958e-01 -1.09686300e-01 -2.52324283e-01 -7.88085535e-02
 -3.87593657e-01 -3.00905436e-01  1.34576499e-01  2.21124947e-01
  1.12909593e-01  8.03473070e-02 -2.31341243e-01 -5.39005280e-01
 -8.75614047e-01  5.43750167e-01  7.40346918e-03  1.77603413e-03
 -1.19095393e-01  2.79089566e-02  3.41629088e-01 -1.25278965e-01
  2.87890017e-01 -3.90708923e-01 -1.75127223e-01 -1.48448246e-02
 -9.28119197e-02 -6.45016506e-02 -1.61496311e-01 -3.76103586e-03
  7.16901124e-02  1.60841838e-01 -1.61558047e-01  1.58174723e-01
 -9.45717916e-02 -1.34893820e-01  1.60557225e-01  2.30733752e-01
  1.32042188e-02 -8.75526853e-03 -1.78873852e-01  1.71281714e-02
 -2.74051905e-01 -3.87778789e-01 -1.37573853e-01  2.72888720e-01
 -1.59498677e-02 -4.32099760e-01 -2.52203250e+00 -1.29231542e-01
 -4.12842780e-01 -3.83452713e-01 -2.11556509e-01  2.28122070e-01
  2.05609605e-01  5.08077323e-01  4.08126116e-01 -2.13666353e-02
 -2.56014109e-